Civi/Api4/Event/Subscriber/ActingUserSubscriber.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Civi\Api4\Event\Subscriber;
+
+use Civi\API\Events;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Set acting user for api request, overriding the current acting user
+ */
+class ActingUserSubscriber implements EventSubscriberInterface {
+
+  /**
+   * @return array
+   */
+  public static function getSubscribedEvents() {
+    return [
+      Events::AUTHORIZE => [
+        ['onApiAuthorize', 500],
+      ],
+      Events::RESPOND => [
+        ['onApiRespond', 500],
+      ],
+      Events::EXCEPTION => [
+        ['onApiRespond', 500],
+      ],
+    ];
+  }
+
+  /**
+   * Change acting user.
+   *
+   * Push current user onto the stack.
+   *
+   * @param \Civi\API\Event\AuthorizeEvent $event
+   */
+  public function onApiAuthorize(\Civi\API\Event\AuthorizeEvent $event) {
+    /* @var \Civi\Api4\Generic\AbstractAction $apiRequest */
+    $apiRequest = $event->getApiRequest();
+    if ($apiRequest['version'] == 4 && $apiRequest->getActingUser()) {
+      \CRM_Core_Session::singleton()->overrideCurrentUser($apiRequest->getActingUser(), $apiRequest['id']);
+      // Set logging contact if this is a write operation
+      if (substr($apiRequest->getActionName(), 0, 3) != 'get') {
+        \CRM_Core_DAO::executeQuery('SET @civicrm_user_id = %1', [1 => [(int) $apiRequest->getActingUser(), 'Integer']]);
```

Comment:
IMHO, it should be an invariant condition that the PHP session value for `userID` always matches the MySQL `@civicrm_user_id`. They should always flip back/forth together.

It also shouldn't matter if the active operation is a `get` or `create`. Why would you care during reads?
* Permissioning - Some current/future SQL queries or customizations could reasonably be using `@civicrm_user_id` as a filter criterion.
* Logging/side-effects - The user may generally think of their request as read-oriented; but some organizations have policies in which reads should be logged, and you can flip a coin on whether that customization would use MySQL's `@civicrm_user_id`  or PHP's `userID`.

Reply:
Ok so you think we should set the `@civicrm_user_id` in core rather than in the api?